File: src/YowThi.Erp.Web/src/features/party/supplierMaster.ts
import { getApiJson, postApiCommand } from '../../app/api/apiTransport';
import type { OperationalLocale } from '../../app/i18n/locale';

export { ApiProblemError } from '../../app/api/apiTransport';

export type SupplierMasterStatus = 'all' | 'active' | 'inactive' | 'deleted';

export interface SupplierMasterItem {
  id: string;
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
  rowVersion: number;
  createdAt: string;
  deletedAt: string | null;
}

export interface SupplierMasterPage {
  items: SupplierMasterItem[];
  nextOffset: number | null;
}

export interface SupplierDraftRequest {
  nameZhTw: string | null;
  nameThTh: string | null;
  bankName: string | null;
  bankAccount: string | null;
  phone: string | null;
  address: string | null;
  active: boolean;
}

export type CreateSupplierRequest = SupplierDraftRequest;

export interface UpdateSupplierRequest extends SupplierDraftRequest {
  expectedRowVersion: number;
}

export interface SupplierWriteResult {
  supplierId: string;
  rowVersion: number;
}

interface SupplierQuery {
  locale: OperationalLocale;
  search?: string;
  status?: SupplierMasterStatus;
  offset?: number;
  limit?: number;
  signal?: AbortSignal;
}

interface SupplierCommandOptions {
  locale: OperationalLocale;
  idempotencyKey: string;
  signal?: AbortSignal;
}

export function listSuppliers(query: SupplierQuery): Promise<SupplierMasterPage> {
  const parameters = new URLSearchParams();
  const search = query.search?.trim();
  if (search) parameters.set('search', search);
  if (query.status) parameters.set('status', query.status);
  if (query.offset !== undefined) parameters.set('offset', String(query.offset));
  if (query.limit !== undefined) parameters.set('limit', String(query.limit));
  const suffix = parameters.size === 0 ? '' : `?${parameters.toString()}`;
  return getApiJson<SupplierMasterPage>(`/api/v1/party/suppliers${suffix}`, {
    locale: query.locale,
    signal: query.signal,
  });
}

export function createSupplier(
  request: CreateSupplierRequest,
  options: SupplierCommandOptions,
): Promise<SupplierWriteResult> {
  return postApiCommand<CreateSupplierRequest, SupplierWriteResult>(
    '/api/v1/party/suppliers',
    request,
    options,
  );
}

export function updateSupplier(
  supplierId: string,
  request: UpdateSupplierRequest,
  options: SupplierCommandOptions,
): Promise<SupplierWriteResult> {
  return postApiCommand<UpdateSupplierRequest, SupplierWriteResult>(
    `/api/v1/party/suppliers/${supplierId}/update`,
    request,
    options,
  );
}

File: src/YowThi.Erp.Application/Party/SupplierMaster.cs
using YowThi.Erp.Application.Common.Errors;
using YowThi.Erp.Application.Common.Idempotency;
using YowThi.Erp.Application.Common.Identity;
using YowThi.Erp.Application.Common.Results;

namespace YowThi.Erp.Application.Party;

public sealed record CreateSupplierCommand(
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    string? Code = null);

public sealed record CreateSupplierExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    CreateSupplierCommand Command);

public sealed record UpdateSupplierCommand(
    Guid SupplierId,
    long ExpectedRowVersion,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    string? Code = null);

public sealed record UpdateSupplierExecution(
    CommandId CommandId,
    CommandRequestHash RequestHash,
    ActorAccountId ActorAccountId,
    UpdateSupplierCommand Command);

public sealed record SupplierMasterWriteResult(
    Guid SupplierId,
    long RowVersion);

public interface ISupplierMasterExecutor
{
    ValueTask<ApplicationResult<SupplierMasterWriteResult>> CreateAsync(
        CreateSupplierExecution execution,
        CancellationToken cancellationToken);

    ValueTask<ApplicationResult<SupplierMasterWriteResult>> UpdateAsync(
        UpdateSupplierExecution execution,
        CancellationToken cancellationToken);
}

public enum SupplierMasterStatusFilter
{
    All,
    Active,
    Inactive,
    Deleted,
}

public sealed record SupplierMasterQuery(
    string? Search,
    SupplierMasterStatusFilter Status,
    int Offset,
    int Limit);

public sealed record SupplierMasterItem(
    Guid Id,
    string? Code,
    string? NameZhTw,
    string? NameThTh,
    string? BankName,
    string? BankAccount,
    string? Phone,
    string? Address,
    bool Active,
    long RowVersion,
    DateTimeOffset CreatedAt,
    DateTimeOffset? DeletedAt);

public sealed record SupplierMasterPage(
    IReadOnlyList<SupplierMasterItem> Items,
    int? NextOffset);

public interface ISupplierMasterReader
{
    ValueTask<SupplierMasterPage> GetAsync(
        SupplierMasterQuery query,
        CancellationToken cancellationToken);
}

public static class SupplierMasterErrorCodes
{
    public const string InvalidInput = "party.supplier.invalid-input";
    public const string SupplierNotFound = "party.supplier.not-found";
    public const string SupplierDeleted = "party.supplier.deleted";
    public const string StaleRowVersion = "party.supplier.stale-row-version";
    public const string IdempotencyKeyReused = "party.supplier.idempotency-key-reused";
}

public static class SupplierMasterValidation
{
    public static ApplicationError? Validate(CreateSupplierCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return HasVisibleName(command.NameZhTw, command.NameThTh)
            ? null
            : Invalid();
    }

    public static ApplicationError? Validate(UpdateSupplierCommand command)
    {
        ArgumentNullException.ThrowIfNull(command);
        return command.SupplierId != Guid.Empty
            && command.ExpectedRowVersion >= 1
            && HasVisibleName(command.NameZhTw, command.NameThTh)
                ? null
                : Invalid();
    }

    private static bool HasVisibleName(string? nameZhTw, string? nameThTh) =>
        !string.IsNullOrWhiteSpace(nameZhTw) || !string.IsNullOrWhiteSpace(nameThTh);

    private static ApplicationError Invalid() =>
        ApplicationError.Create(
            ApplicationErrorKind.Validation,
            SupplierMasterErrorCodes.InvalidInput);
}
